utils/Classes/Panel.js

```diff
@@ -22,136 +15,194 @@ export default class Panel {
+
+        if (parsedState.elements.size > 0) {
+            let elementsToDelete = new Set(this.elements.keys());
+            parsedState.elements.forEach((value, key) => {
+                elementsToDelete.delete(key);
+                if (!this.hasElement(key)) {
+                    let tempType =
+                        value.type._name === 'Asset'
+                            ? new Asset(value.type._path)
+                            : new Text(value.type._content, value.type._fontSize, value.type.fontFamily);
+                    let tempEl = new ElementDS(value.width, value.height, value.alt, tempType);
+                    tempEl.id = value.id;
+                    tempEl.pos = new Position(value.pos._x, value.pos._y);
+                    tempEl.isMirroredHorizontal = value.isMirroredHorizontal;
+                    tempEl.isMirroredVertical = value.isMirroredVertical;
+                    tempEl.rotation = value.rotation;
+                    tempEl.z = value.z;
+                    tempEl.width = value.width;
+                    tempEl.height = value.height;
+                    this.elements.set(tempEl.id, tempEl);
+                } else {
+                    let targetElement = this.getElement(key);
+                    this.setPropertiesTo(
+                        value,
+                        targetElement,
+                        ['width', 'height', 'pos', 'isMirroredHorizontal', 'isMirroredVertical', 'rotation', 'z'],
+                        true
+                    );
+                }
+            });
+
+            // Remove elements that are not present in the new state
+            elementsToDelete.forEach((key) => {
+                this.elements.delete(key);
+            });
+        } else {
+            this.clearMaps();
+        }
+    }
+
+    setPropertiesTo(oSource, oTarget, properties, useDashPrefix) {
+        if (oTarget._type._name === 'Text') {
+            oTarget._type._content = oSource.type._content;
+            oTarget._type._fontSize = oSource.type._fontSize;
+            oTarget._type._fontFamily = oSource.type._fontFamily;
+        }
+        properties.forEach((property) => {
+            let propertyName = useDashPrefix ? `_${property}` : property;
+            if (!oSource.hasOwnProperty(property)) return;
+            if (property === 'pos' && oSource[property]._x !== undefined && oSource[property]._y !== undefined) {
+                oTarget[propertyName] = new Position(oSource[property]._x, oSource[property]._y);
+            } else if (oTarget.hasOwnProperty(propertyName)) {
+                oTarget[propertyName] = oSource[property];
+            }
+        });
+    }
+
+    addAlteration() {
+        this._redo = [];
+        let currentState = this.toJSON();
+        this.history.push(currentState);
+        if (this.history.length > this.maxHistoryLength) {
+            this.history.shift();
+        }
+    }
+
+    undo() {
+        if (this.history.length <= 1) return;
```

Comment:
I don't think you need to control the history length here. This function can only be called when the button is enabled (which can only occur if history has something in it)